### ISSUE TITLE
Wrong formatting for '&&' after qualified constants in expressions

### ISSUE THREAD

**Issue by f7d9a599-8365-4225-8735-d1e6340786a8** [llvm/llvm-project#47209]
|  |  |
| --- | --- |
| Bugzilla Link | [47865](https://llvm.org/bz47865) |
| Version | trunk |
| OS | Linux |

## Extended Description 
Example:

    enum class A: int { X = 0, Y, Z };
    template <typename T, A Size = static_cast<A>(sizeof(T))>
    inline constexpr bool ah = Size != A::X && Size != A::Y;

The inline constexpr got formatted as:

    inline constexpr bool ah = Size != A::X &&Size != A::Y;

As if A::X is a type, and Size is a rvalue reference.

**Comment by mkurdej** [comment]
Apparently the bug is because of the expression being misparsed in a template (inline variable) and so `&&` gets recognized as `PointerOrReference`.
With LLVM style:
```
bool v = 0 != A && B;               // ok
template <> bool v = 0 != A &&B;    // BAD
template <> bool v = 0 != 1 && B;   // ok
template <> bool v = (0 != A) && B; // ok
```

Good:
```
AnnotatedTokens(L=0):
 M=0 C=0 T=Unknown S=1 F=0 B=0 BK=0 P=0 Name=bool L=4 PPK=2 FakeLParens=2/ FakeRParens=0 II=0x1739b5cc9c8 Text='bool'
 M=0 C=1 T=StartOfName S=1 F=0 B=0 BK=0 P=220 Name=identifier L=6 PPK=2 FakeLParens= FakeRParens=0 II=0x1739b6018e8 Text='v'
 M=0 C=0 T=BinaryOperator S=1 F=0 B=0 BK=0 P=22 Name=equal L=8 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text='='
 M=0 C=1 T=Unknown S=1 F=0 B=0 BK=0 P=22 Name=numeric_constant L=10 PPK=2 FakeLParens=9/5/ FakeRParens=0 II=0x0 Text='0'
 M=0 C=0 T=BinaryOperator S=1 F=0 B=0 BK=0 P=29 Name=exclaimequal L=13 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text='!='
 M=0 C=1 T=Unknown S=1 F=0 B=0 BK=0 P=29 Name=identifier L=15 PPK=2 FakeLParens= FakeRParens=1 II=0x1739b601918 Text='A'
 M=0 C=0 T=BinaryOperator S=1 F=0 B=0 BK=0 P=25 Name=ampamp L=18 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text='&&'
 M=0 C=1 T=Unknown S=1 F=0 B=0 BK=0 P=25 Name=identifier L=20 PPK=2 FakeLParens= FakeRParens=2 II=0x1739b601948 Text='B'
 M=0 C=0 T=Unknown S=0 F=0 B=0 BK=0 P=23 Name=semi L=21 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text=';'
----
AnnotatedTokens(L=0):
 M=0 C=0 T=Unknown S=1 F=0 B=0 BK=0 P=0 Name=eof L=0 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text=''
----
bool v = 0 != A && B;
```

Bad:
```
AnnotatedTokens(L=0):
 M=0 C=0 T=Unknown S=1 F=0 B=0 BK=0 P=0 Name=template L=8 PPK=2 FakeLParens=2/ FakeRParens=0 II=0x1e923492cc8 Text='template'
 M=0 C=0 T=TemplateOpener S=1 F=0 B=0 BK=0 P=30 Name=less L=10 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text='<'
 M=0 C=0 T=TemplateCloser S=0 F=0 B=0 BK=0 P=360 Name=greater L=11 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text='>'
 M=0 C=1 T=Unknown S=1 F=0 B=0 BK=0 P=30 Name=bool L=16 PPK=2 FakeLParens= FakeRParens=0 II=0x1e9234928e8 Text='bool'
 M=0 C=1 T=StartOfName S=1 F=0 B=0 BK=0 P=220 Name=identifier L=18 PPK=2 FakeLParens= FakeRParens=0 II=0x1e9234d0a68 Text='v'
 M=0 C=0 T=BinaryOperator S=1 F=0 B=0 BK=0 P=22 Name=equal L=20 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text='='
 M=0 C=1 T=Unknown S=1 F=0 B=0 BK=0 P=22 Name=numeric_constant L=22 PPK=2 FakeLParens=9/ FakeRParens=0 II=0x0 Text='0'
 M=0 C=0 T=BinaryOperator S=1 F=0 B=0 BK=0 P=29 Name=exclaimequal L=25 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text='!='
 M=0 C=1 T=Unknown S=1 F=0 B=0 BK=0 P=29 Name=identifier L=27 PPK=2 FakeLParens= FakeRParens=0 II=0x1e9234d0a98 Text='A'
 M=0 C=1 T=PointerOrReference S=1 F=0 B=0 BK=0 P=210 Name=ampamp L=30 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text='&&'
           ^^^^^^^^^^^^^^^^^^
 M=0 C=1 T=StartOfName S=0 F=0 B=0 BK=0 P=220 Name=identifier L=31 PPK=2 FakeLParens= FakeRParens=2 II=0x1e9234d0ac8 Text='B'
           ^^^^^^^^^^^
 M=0 C=0 T=Unknown S=0 F=0 B=0 BK=0 P=23 Name=semi L=32 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text=';'
----
AnnotatedTokens(L=0):
 M=0 C=0 T=Unknown S=1 F=0 B=0 BK=0 P=0 Name=eof L=0 PPK=2 FakeLParens= FakeRParens=0 II=0x0 Text=''
----
template <> bool v = 0 != A &&B;
```

**Comment by rymiel** [comment]
This appears to be fixed by 3227aa3aa83440ff94a3b13a29623e03b05098f2
